README
This is the GEANT4 model of the JLab 5 MeV Mott Polarimeter. Any questions should be referred to 
Martin McHugh at mjmchugh@jlab.org.


Mott_Polarimeter/History
l$Id: History,v 1.64 2010-06-06 04:05:29 perl Exp $
-------------------------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    Example N02 History file
                    ------------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

2013-03-09 Martin McHugh (MottDetectorPackage-V01-00-00)
- Copied files from exampleN02 changed names appropriately.
     
12-12-11 mma (exampleN02-V09-05-00)
- modify SteppingVerbose for OutOfWorld
     
29-11-11 Ben Morgan (exampleN02-V09-04-04)
- CMakeLists.txt edited to add standard UI/Vis activation and copying of scripts
  to build directory, plus comments and neatification.

15-11-11 G.Cosmo (exampleN02-V09-04-03)
- Get rid of call to obsolete G4Exception signature in
  ExN02ChamberParameterisation.cc.

03-06-11 mma (exampleN02-V09-04-02)     
- PhysicsList migrated to new scheme G4PhysListHelper.
     
10-02-11 mma (exampleN02-V09-04-01)
- Updated SteppingVerbose.

01-02-11 mma (exampleN02-V09-04-00)
- /vis/modeling/trajectories/drawByCharge-0/default/setStepPtsSize 1.

04-06-10 J.Perl (exampleN02-V09-03-01)
- Updated vis usage.
    
22-01-10 mma (exampleN02-V09-03-00)
- DetectorConstruction::SetMagField() : fpMagField->SetMagFieldValue().
   
15-11-09 mma (exampleN02-V09-02-01)
- Use G4MuMultipleScattering for muons.
- Separated alpha from GenericIon.
    
30 October 2009 John Allison (exampleN02-V09-02-00)
- Introduced G4UIExecutive.     

17-12-08 mma
- vis.mac : /vis/scene/add/trajectories smooth.

31-10-08 mma (exampleN02-V09-01-04)
- Use G4hMultipleScattering for muons and protons.
     
26-10-08 mma (exampleN02-V09-01-03)
- Updated PhysicsList for proton and ions.
     
22-09-08 mma (exampleN02-V09-01-02)
- PhysicsList : AddStepMax() for G4StepLimiter.
- DetectorConstruction : SetMaxStep(value).
- DetectorMessenger : /N02/det/stepMax.
     
07-05-08 J.Allison (exampleN02-V09-01-01)
- Protected "/control/execute vis.mac" with G4VIS_USE flag.
     
03-05-08 mma (exampleN02-V09-01-00)
- PhysicsList : replace G4MultipleScattering by G4eMultipleScattering
     
17-10-07 G.Cosmo (exampleN02-V09-00-02)
- Explicitely delete allocated parameterisation and step-limit objects
  in ExN02DetectorConstruction class.

22-08-07 G.Cosmo (exampleN02-V09-00-01)
- Fixed implementation of ExN02MagneticField class in way field is set...

25-06-07 M.Maire (exampleN02-V08-03-01, exampleN02-V09-00-00)
- Cosmetics in main().

11-05-07 G.Cosmo (exampleN02-V08-03-00)
- Show usage of G4GeometryTolerance tuned for the geometry topology.

12-12-06 G.Cosmo (exampleN02-V08-01-01)
- Commented out drawing of hits in vis.mac.

06-11-06 J.Allison (exampleN02-V08-01-00)
- Improved vis.mac to vis best practice.

16-06-06 G.Cosmo (exampleN02-V08-00-03)
- Use coherent allocation scheme for user-classes and
  initialisation in main().

15-06-06 G.Cosmo (exampleN02-V08-00-02)
- Separate instantiation of the user-stepping-verbose class from
  initialisation of the G4VSteppingVerbose singleton.

02-06-06 mma (exampleN02-V08-00-01)
- ExN02TrackerSD.cc: replace Touchable by TouchableHandle
          
31-03-06 mma (exampleN02-V08-00-00)
- Add a sentence in README for G4UserLimits and G4StepLimiter.
     
01-06-05 J.Allison  (exampleN02-V07-00-03)
- ExN02TrackerHit.cc: Changed screen size to 2 pixels.
     
31-05-05 mma (exampleN02-V07-00-02)
- G4UserLimits + G4StepLimiter become uncommented
     
30-05-05  J.Allison  (exampleN02-V07-00-01)
- Completed replacement of vis manager with G4VisExecutive.

3rd May 2005 J.Allison  (examples-V07-00-03)
- Replaced vis manager with G4VisExecutive.

28-01-05 mma (exampleN02-V07-00-00)
- PhysicsList: add G4StepLimiter (in a commented line)
     
16-09-04 mma (exampleN02-V06-02-01)
- DetectorConstruction: modify a comment in PVPlacement. // copy number
     
08-09-04 J.Allison (exampleN02-V06-02-00)
- Removed trajectory printing from ExN02EventAction - now available
  with /vis/scene/add/trajectories, which is now included in vis.mac.

05-04-04 mma (exampleN02-V06-01-00)
- ExN02DetectorConstruction: correct definition of XenonGas
     
19-03-04 mma (exampleN02-V06-00-01)
- ExN02PhysicsList: cleanup of include
     
08-03-04 mma (exampleN02-V06-00-00)
- ExN02RunAction: removed ApplyCommand("/vis/...") statements.

10-11-03 G.Cosmo (exampleN02-V05-02-03)
- Removed OPACS driver from vis-manager.
- Cleared warnings for parameterised detectors.

30-10-03 G.Cosmo (exampleN02-V05-02-02)
- Cleared compilation warnings due to new class G4Orb.

24-10-03 mma (exampleN02-V05-02-01)
- PhysicsList: AddProcess(Bremsstrahlung,-1,3,3) ..etc..
     
06-10-03 mma (exampleN02-V05-02-00)
- cosmetic cleanup in material definitions

28-05-03 G.Cosmo (exampleN02-V05-01-00)
- Get rid of pedantic warnings...

25-03-03 mma (exampleN02-V05-00-02)
- Use G4PVPlacement and G4PVParametrised constructors in logical mother. 
    
11-02-03 mma (exampleN02-V05-00-00)
- a blank in SteppingVerbose
     
27-11-01 mma (exampleN02-V03-02-04)
- vis.mac: /vis/viewer/set/viewpointThetaPhi

19-11-01 J.Allison (exampleN02-V03-02-03)
- Updated ExN02VisManager.cc for new graphics systems and defaults.

07-11-01 S.Tanaka (exampleN02-V03-02-02)
- vis.mac: /vis/viewer/refresh + /vis/viewer/update is replaced 
  with their equivalent command /vis/viewer/flush. 
     
24-10-01 mma (exampleN02-V03-02-01)
- initInter.mac renamed vis.mac
     
19-10-01 S.O'Neale (examples-V03-02-00)
- Update reference output

11-10-01 mma (exampleN02-V03-02-00)

Mott_Polarimeter/src/MottRunAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
// $Id: MottRunAction.cc,v 3.9 2013/12/05 19:56:26 mjmchugh Exp $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "MottRunAction.hh"
#include "MottAnalysis.hh"

#include "G4Run.hh"
#include "G4RunManager.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4ios.hh"

#include <sstream>

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

MottRunAction::MottRunAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

MottRunAction::~MottRunAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void MottRunAction::BeginOfRunAction(const G4Run* aRun)
{

  G4int runID = aRun->GetRunID();
  std::stringstream runNo;
  runNo << runID;
  
  G4cout << "### Run " << runID << " start." << G4endl;
  
  // Create analysis manager
  // The choice of analysis technology is done via selectin of a namespace
  // in MottAnalysis.hh
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  G4cout << "Using " << analysisManager->GetType() 
         << " analysis manager" << G4endl;

  // Open an output ROOTfile
  // G4String fileName = "MottSim_" + runNo.str() + ".root";
  // For the farm
  G4String fileName = "/volatile/hallc/qweak/mjmchugh/Jan_2014/MottSim_" + runNo.str() + ".root";
  analysisManager->OpenFile(fileName);
  analysisManager->SetFirstHistoId(1);
  
  // Creating Ntuple
  analysisManager->CreateNtuple("Mott", "Edep and TrackL");
  
  // UP Detector (0-3)
  analysisManager->CreateNtupleDColumn("up_e");
  analysisManager->CreateNtupleDColumn("up_de");
  analysisManager->CreateNtupleDColumn("up_e_dl");
  analysisManager->CreateNtupleDColumn("up_de_dl");
  
  // DOWN Detector (4-7)
  analysisManager->CreateNtupleDColumn("down_e");
  analysisManager->CreateNtupleDColumn("down_de");
  analysisManager->CreateNtupleDColumn("down_e_dl");
  analysisManager->CreateNtupleDColumn("down_de_dl");
  
  // LEFT Detector (8-11)
  analysisManager->CreateNtupleDColumn("left_e");
  analysisManager->CreateNtupleDColumn("left_de");
  analysisManager->CreateNtupleDColumn("left_e_dl");
  analysisManager->CreateNtupleDColumn("left_de_dl");
  
  // RIGHT Detector (12-15)
  analysisManager->CreateNtupleDColumn("right_e");
  analysisManager->CreateNtupleDColumn("right_de");
  analysisManager->CreateNtupleDColumn("right_e_dl");
  analysisManager->CreateNtupleDColumn("right_de_dl");
  
  // "PMT" response.
  analysisManager->CreateNtupleIColumn("Up_E_PMT");	// 16
  analysisManager->CreateNtupleIColumn("Up_dE_PMT");	// 17
  analysisManager->CreateNtupleIColumn("Down_E_PMT");	// 18
  analysisManager->CreateNtupleIColumn("Down_dE_PMT");  // 19
  analysisManager->CreateNtupleIColumn("Left_E_PMT");   // 20
  analysisManager->CreateNtupleIColumn("Left_dE_PMT");  // 21
  analysisManager->CreateNtupleIColumn("Right_E_PMT");  // 22
  analysisManager->CreateNtupleIColumn("Right_dE_PMT"); // 23
  
  // Event #
  analysisManager->CreateNtupleIColumn("Event_ID");     // 24  
  
  // Dump segments.
  G4int nBeSegments = 20*6;
  G4int nCuSegments = 20*18;
  for(G4int i=0; i<nBeSegments; i++) {
    G4String name = "BeSegment";
    std::ostringstream number;
    number << i;
    name += number.str();
    analysisManager->CreateNtupleDColumn(name);
  }
  for(G4int i=0; i<nCuSegments; i++) {
    G4String name = "CuSegment";
    std::ostringstream number;
    number << i;
    name += number.str();
    analysisManager->CreateNtupleDColumn(name);
  }
  
  
  analysisManager->FinishNtuple();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void MottRunAction::EndOfRunAction(const G4Run*)
{
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();

  // Write and Save ROOTfile.
  analysisManager->Write();
  analysisManager->CloseFile(); 

  // Complete cleanup
  delete G4AnalysisManager::Instance();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......



